### PR TITLE
Adding methods to check read/write access in Image

### PR DESCRIPTION
**Pull request by jjbustamante**
Signed-off-by: Juan Bustamante <jbustamante@vmware.com>

## Summary

This PR is adding a couple of methods in the Image object to verify if the client has read/write access in the registry, these methods will be used to fix the issue [572 - validate registry credentials](https://github.com/buildpacks/lifecycle/issues/572) in the [lifecycle](https://github.com/buildpacks/lifecycle) repository

## Documentation

- Should this change be documented?
    - [x] No

## Related
Resolves #572 (in lifecycle)

### CODE REVIEW

File: remote/remote.go [hunk 1]
```diff
@@ -347,12 +347,36 @@ func (i *Image) Name() string {
...
+		if transportErr, ok := err.(*transport.Error); ok {
+			return transportErr.StatusCode != http.StatusUnauthorized &&
+				transportErr.StatusCode != http.StatusForbidden
+		}
+		return false
+	}
+	return true
+}
+
+func (i *Image) headManifest() (*v1.Descriptor, error) {
```
Comment (aemengo) [inline remote/remote.go:374]:
```suggestion
func (i *Image) found() (*v1.Descriptor, error) {
```

Love all this! Let's rename this to lowercase `found()`.
Having a similarly named **Uppercase/lowercase** methods is a (completely optional) golang idiom in these types of cases.

File: remote/remote_test.go [hunk 2]
```diff
@@ -1500,4 +1500,82 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
...
+			it("returns false", func() {
+				image, err := remote.NewImage(repoName, authn.DefaultKeychain)
+				h.AssertNil(t, err)
+				h.AssertEq(t, image.CheckReadAccess(), false)
+			})
+		})
+	})
+
+	when("#CheckReadWriteAccess", func() {
+		when("image exists in the registry and client has read access", func() {
```
Comment (natalieparellano) [inline remote/remote_test.go:1544]:
```suggestion
		when("image exists in the registry and client has read/write access", func() {
```

Comment (natalieparellano) [reply remote/remote_test.go:1544]: (is this what you mean?)

File: remote/remote_test.go [hunk 3]
```diff
@@ -1500,4 +1500,82 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
...
+
+		when("image does not exist in the registry and client has read access", func() {
+			it("returns true", func() {
+				image, err := remote.NewImage(repoName, authn.DefaultKeychain)
+				h.AssertNil(t, err)
+				h.AssertEq(t, image.CheckReadWriteAccess(), true)
+			})
+		})
+
+		when("image does not exist in the registry and client hasn't read access", func() {
```
Comment (natalieparellano) [inline remote/remote_test.go:1565]:
```suggestion
		when("image does not exist in the registry and client doesn't have read access", func() {
```

Comment (natalieparellano) [reply remote/remote_test.go:1565]: Is there a way to make a "read only" test fixture?

File: remote/remote_test.go [hunk 4]
```diff
@@ -1500,4 +1500,82 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
...
+			it("returns true", func() {
+				image, err := remote.NewImage(repoName, authn.DefaultKeychain)
+				h.AssertNil(t, err)
+				h.AssertEq(t, image.CheckReadAccess(), true)
+			})
+		})
+
+		when("image does not exist in the registry and client hasn't read access", func() {
+			it.Before(func() {
+				os.Setenv("DOCKER_CONFIG", readonlyDockerRegistry.DockerDirectory)
```
Comment (natalieparellano) [inline remote/remote_test.go:1528]: Maybe this is just the variable name, but I find this confusing - shouldn't the client still have read access to the "read only" registry?

File: remote/remote_test.go [hunk 5]
```diff
@@ -1500,4 +1500,82 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
...
+
+		when("image does not exist in the registry and client has read access", func() {
+			it("returns true", func() {
+				image, err := remote.NewImage(repoName, authn.DefaultKeychain)
+				h.AssertNil(t, err)
+				h.AssertEq(t, image.CheckReadAccess(), true)
+			})
+		})
+
+		when("image does not exist in the registry and client hasn't read access", func() {
```
Comment (natalieparellano) [inline remote/remote_test.go:1526]:
```suggestion
		when("image does not exist in the registry and client doesn't have read access", func() {
```

File: remote/remote_test.go [hunk 6]
```diff
@@ -1500,4 +1500,82 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
...
+				h.AssertNil(t, origImage.Save())
+
+				image, err := remote.NewImage(repoName, authn.DefaultKeychain)
+				h.AssertNil(t, err)
+
+				h.AssertEq(t, image.CheckReadWriteAccess(), true)
+			})
+		})
+
+		when("image does not exist in the registry and client has read access", func() {
```
Comment (natalieparellano) [inline remote/remote_test.go:1557]:
```suggestion
		when("image does not exist in the registry and client has read/write access", func() {
```

Comment (natalieparellano) [reply remote/remote_test.go:1557]: Or maybe "only has read access" as mentioned here: https://github.com/buildpacks/imgutil/pull/123/files#r646862762

File: remote/remote_test.go (hunk 4 above)
Comment (jjbustamante) [reply remote/remote_test.go:1528]: When I switched to the read-only registry my auth.DefaultKeyChain is still pointing to dockerRegistry user/pass configuration. in this case, I am trying to access an image that doesn't exist in the registry with invalid credentials, that's why I do not have read access. 

I agreed, maybe the variable name is not appropriated, but it was already there, I didn't want to create another registry instance. Maybe I can point to the same registry with another variable name? could it make more clear?

File: remote/remote.go (hunk 1 above)
Comment (jjbustamante) [reply remote/remote.go]: I addressed this suggestion in the following [commit](https://github.com/buildpacks/imgutil/pull/123/commits/d588cd706db0e32f55858fa00211c948d045210a)

File: remote/remote_test.go (hunk 2 above)
Comment (jjbustamante) [reply remote/remote_test.go]: I addressed this suggestion in the following [commit](https://github.com/buildpacks/imgutil/pull/123/commits/d588cd706db0e32f55858fa00211c948d045210a)

File: remote/remote_test.go (hunk 5 above)
Comment (jjbustamante) [reply remote/remote_test.go]: I addressed this suggestion in the following [commit](https://github.com/buildpacks/imgutil/pull/123/commits/d588cd706db0e32f55858fa00211c948d045210a)

File: remote/remote_test.go (hunk 4 above)
Comment (natalieparellano) [reply remote/remote_test.go:1528]: How about just `os.Unsetenv`?

Comment (jjbustamante) [reply remote/remote_test.go:1528]: Oh! that one seems to be even better!!! Thanks!!!

File: remote/remote_test.go (hunk 3 above)
Comment (jjbustamante) [reply remote/remote_test.go]: Done! added in the following [commit](https://github.com/buildpacks/imgutil/pull/123/commits/03a54599773fab9cb4690816c4dbe72252711908)

File: remote/remote_test.go (hunk 4 above)
Comment (jjbustamante) [reply remote/remote_test.go]: > How about just `os.Unsetenv`?

I replaced the `it.Before` with the `os.Unsetenv` in the following [commit](https://github.com/buildpacks/imgutil/pull/123/commits/03a54599773fab9cb4690816c4dbe72252711908)

File: remote/remote_test.go [hunk 7]
```diff
@@ -1500,4 +1500,100 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
...
+			})
+		})
+
+		when("image does not exist in the registry and client has read/write access", func() {
+			it("returns true", func() {
+				image, err := remote.NewImage(repoName, authn.DefaultKeychain)
+				h.AssertNil(t, err)
+				h.AssertEq(t, image.CheckReadWriteAccess(), true)
+			})
+		})
```
Comment (natalieparellano) [inline remote/remote_test.go:1581]: This seems to be the most valuable "does not exist in the registry" case. I think the other "does not exist in the registry" cases might make more sense as "exists in the registry" cases. But maybe that's just me...

File: remote/remote_test.go [hunk 8]
```diff
@@ -1500,4 +1500,100 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
...
+			})
+
+			it("returns true", func() {
+				image, err := remote.NewImage(repoName, authn.DefaultKeychain)
+				h.AssertNil(t, err)
+				h.AssertEq(t, image.CheckReadWriteAccess(), true)
+			})
+		})
+
+		when("image does not exist in the registry and client only has read access", func() {
```
Comment (natalieparellano) [inline remote/remote_test.go:1559]: See https://github.com/buildpacks/imgutil/pull/123/files#r648687998

File: remote/remote_test.go [hunk 9]
```diff
@@ -1500,4 +1500,100 @@ func testImage(t *testing.T, when spec.G, it spec.S) {
...
+
+		when("image does not exist in the registry and client has read/write access", func() {
+			it("returns true", func() {
+				image, err := remote.NewImage(repoName, authn.DefaultKeychain)
+				h.AssertNil(t, err)
+				h.AssertEq(t, image.CheckReadWriteAccess(), true)
+			})
+		})
+
+		when("image does not exist in the registry and client doesn't have read/write access", func() {
```
Comment (natalieparellano) [inline remote/remote_test.go:1583]: See https://github.com/buildpacks/imgutil/pull/123/files#r648687998

**Review by natalieparellano — APPROVED** [review]
This looks great! I left a comment about inverting a couple "does not exist in the registry" cases (for readability), but it's just a suggestion.

**Pull request merged**